server/src/main/java/org/elasticsearch/cluster/routing/allocation/AbstractAllocationDecision.java

```diff
@@ -110,6 +111,13 @@ public static XContentBuilder discoveryNodeToXContent(DiscoveryNode node, boolea
             }
             builder.endObject();
         }
+        if (node.getRoles().isEmpty() == false) {
```

Comment:
In practice I think the roles will never be empty here, we only ever report allocation decisions for data nodes, but it makes the client developers' lives easier if we expose a fixed set of fields as far as possible. Let's not check for emptiness here and always emit `roles`. (I know `attributes` appears and disappears like this, but that's been like that forever).

Reply:
Yes that's right.